src/editors/containers/TextEditor/components/ImageSettingsModal/hooks.js

```diff
@@ -0,0 +1,251 @@
+    : { changed: dimKeys.width, other: dimKeys.height };
+
+  const direction = local[keys.changed] > dimensions[keys.changed] ? 1 : -1;
+
+  // don't move down if already at minimum size
+  if (direction < 0 && isMin) { return dimensions; }
+  // find closest valid iteration of the changed field
+  iter = Math.max(Math.round(local[keys.changed] / minInc[keys.changed]), 1);
+  // if closest valid iteration is current iteration, move one iteration in the change direction
+  if (iter === (dimensions[keys.changed] / minInc[keys.changed])) { iter += direction; }
+
+  out[keys.changed] = iter * minInc[keys.changed];
+  out[keys.other] = out[keys.changed] * (locked[keys.other] / locked[keys.changed]);
+
+  return out;
+};
+
+/**
+ * newDimensions({ dimensions, local, locked })
+ * Returns the local dimensions if unlocked or unchanged, and otherwise returns new valid
+ * dimensions.
+ * @param {obj} dimensions - current stored dimensions
+ * @param {obj} local - local (active) dimensions in the inputs
+ * @param {obj} locked - locked dimensions
+ * @return {obj} - output dimensions after attempted move ({ height, width })
+ */
+export const newDimensions = ({ dimensions, local, locked }) => (
+  (!locked || checkEqual(local, dimensions))
+    ? local
+    : module.getValidDimensions({ dimensions, local, locked })
+);
+
+/**
+ * lockDimensions({ dimensions, lockInitialized, setLocked })
+ * Lock dimensions if lock initialized.  Store minimum valid increment on lock so
+ * that we don't have re-compute.
+ * @param {obj} dimensions - current stored dimensions
+ * @param {bool} lockInitialized - has the lock state initialized?
+ * @param {func} setLocked - set lock state
+ */
+export const lockDimensions = ({ dimensions, lockInitialized, setLocked }) => {
+  if (!lockInitialized) { return; }
+
+  // find minimum viable increment
+  let gcd = findGcd(dimensions.width, dimensions.height);
+  if ([dimensions.width, dimensions.height].some(v => !Number.isInteger(v / gcd))) {
+    gcd = 1;
+  }
+  const minInc = { width: dimensions.width / gcd, height: dimensions.height / gcd, gcd };
+  setLocked({ ...dimensions, minInc });
+};
+
+/**
+ * dimensionLockHooks({ dimensions })
+ * Returns a set of hooks pertaining to the dimension locks.
+ * Locks the dimensions initially, on lock initialization.
+ * @param {obj} dimensions - current stored dimensions
+ * @return {obj} - dimension lock hooks
+ *   {func} initializeLock - enable the lock mechanism
+ *   {obj} locked - current locked state
+ *   {func} lock - lock the current dimensions
+ *   {func} unlock - unlock the dimensions
+ */
+export const dimensionLockHooks = ({ dimensions }) => {
+  const [locked, setLocked] = module.state.locked(null);
+  const [lockInitialized, setLockInitialized] = module.state.lockInitialized(null);
+  const lock = () => module.lockDimensions({ lockInitialized, dimensions, setLocked });
+
+  React.useEffect(lock, [lockInitialized]);
```

Comment:
Just to clarify-- here, whenever `lockInitialized` changes, the lock function is called to set the dimensions to whatever they were the first time the lock was initialized, or is dimensionLockHooks also called every render, so that this value would lock to the current dimensions. I know it is most likely the former, but want to be clear on our approach.

Reply:
whenever lockInitialized changes, it locks to the "current" dimensions, iff (if-and-only-if) the lock is initialized.
So, in practice, it gets run once, when the lock initializes the first time (and the component renders) and loads the initial dimension values.

Reply:
Ah ok, that makes sense. Thanks for all the additional clarification.